benchmarks/mpi_p2p_bi_avail_mt.cpp

```diff
@@ -0,0 +1,226 @@
+        }
+
+        int dbg = 0, sdbg = 0, rdbg = 0, j;
+        int lsent = 0, lrecv = 0;
+#ifdef USE_TESTANY
+        int flag;
+#endif
+        char header[256];
+        snprintf(header, 256, "%d total bwdt ", rank);
+        while (sent < niter || received < niter)
+        {
+            if (thrid == 0 && sdbg >= (niter / 10))
+            {
+                std::cout << rank << "   " << sent << " sent\n";
+                sdbg = 0;
+            }
+
+            if (thrid == 0 && rdbg >= (niter / 10))
+            {
+                std::cout << rank << "   " << received << " received\n";
+                rdbg = 0;
+            }
+
+            if (thrid == 0 && dbg >= (2 * niter / 10))
+            {
+                dbg = 0;
+                t0.vtoc(header,
+                    (double)(received - last_received + sent - last_sent) * size * buff_size / 2);
+                t0.tic();
+                last_received = received;
+                last_sent = sent;
+            }
+
+#ifdef USE_TESTANY
+            MPI_Testany(inflight, rreq, &j, &flag, MPI_STATUS_IGNORE);
+            if (flag)
+            {
+                MPI_Irecv(rbuffers[j], buff_size, MPI_BYTE, peer_rank, thrid * inflight + j,
+                    mpi_comm, &rreq[j]);
+                dbg += nthr;
+                rdbg += nthr;
+                received++;
+                lrecv++;
+            }
+
+            if (lsent < lrecv + 2 * inflight && sent < niter)
+            {
+                MPI_Testany(inflight, sreq, &j, &flag, MPI_STATUS_IGNORE);
+                if (flag)
+                {
+                    MPI_Isend(sbuffers[j], buff_size, MPI_BYTE, peer_rank, thrid * inflight + j,
+                        mpi_comm, &sreq[j]);
+                    dbg += nthr;
+                    sdbg += nthr;
+                    sent++;
+                    lsent++;
+                }
+            }
+#else
+            for (int i = 0; i < inflight; i++)
+            {
+                MPI_Test(&rreq[i], &j, MPI_STATUS_IGNORE);
+                if (j)
+                {
+                    MPI_Irecv(rbuffers[i], buff_size, MPI_BYTE, peer_rank, thrid * inflight + i,
+                        mpi_comm, &rreq[i]);
+                    dbg += nthr;
+                    rdbg += nthr;
+                    received++;
+                    lrecv++;
+                }
+            }
+            if (lsent < lrecv + 2 * inflight)
+            {
+                for (int i = 0; i < inflight; i++)
+                {
+                    MPI_Test(&sreq[i], &j, MPI_STATUS_IGNORE);
+                    if (j)
+                    {
+                        MPI_Isend(sbuffers[i], buff_size, MPI_BYTE, peer_rank, thrid * inflight + i,
+                            mpi_comm, &sreq[i]);
+                        dbg += nthr;
+                        sdbg += nthr;
+                        sent++;
+                        lsent++;
+                    }
+                }
+            }
+#endif
```

Comment:
can we factor out the body of the loops in a function f.ex. in order not to duplicate code?

Reply:
@boeschf Ha! made a lambda!